From 10d97f25b1ea6c67ef90e838d85cba3171056c26 Mon Sep 17 00:00:00 2001
From: Gu, Wangyi <wangyi.gu@intel.com>
Date: Thu, 2 May 2013 16:45:24 +0800
Subject: make one xvid clip work normal in gallery

BZ: 83021

root cause: the first video sample with MPEG4 codec is a fake sample, which size
is 0, we just skip this kind of video sample, the original design can't handle this
case well.

Category: aosp improvement
Domain: Video.Media-container
Origin: internal
Upstream-Candidate: yes

Change-Id: I8df3a888c5d3b58a9819543f417a291bb493aa5e
Orig-Change-Id: I22b5a1c4d090b49b901f3cd07e1467a86c801429
Signed-off-by: Gu, Wangyi <wangyi.gu@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
Signed-off-by: Yuanjun Huang <yuanjun.huang@intel.com>
---
 media/libstagefright/AVIExtractor.cpp |   21 +++++++++++++++++----
 1 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/media/libstagefright/AVIExtractor.cpp b/media/libstagefright/AVIExtractor.cpp
index 89d1e9f..2d89aa6 100644
--- a/media/libstagefright/AVIExtractor.cpp
+++ b/media/libstagefright/AVIExtractor.cpp
@@ -1017,11 +1017,24 @@ status_t AVIExtractor::addMPEG4CodecSpecificData(size_t trackIndex) {
     size_t size;
     bool isKey;
     int64_t timeUs;
-    status_t err =
-        getSampleInfo(trackIndex, 0, &offset, &size, &isKey, &timeUs);
 
-    if (err != OK) {
-        return err;
+    // Extract codec specific data from the first non-empty sample.
+
+    size_t sampleIndex = 0;
+    for (;;) {
+        status_t err =
+            getSampleInfo(
+                    trackIndex, sampleIndex, &offset, &size, &isKey, &timeUs);
+
+        if (err != OK) {
+            return err;
+        }
+
+        if (size > 0) {
+            break;
+        }
+
+        ++sampleIndex;
     }
 
     sp<ABuffer> buffer = new ABuffer(size);
-- 
1.7.4.1

